src/region.rs

```diff
@@ -65,29 +65,29 @@ impl TryFrom<RegionOrEndpoint> for Region {
 /// Translate an endpoint URL into a Region
 pub fn region_from_endpoint(endpoint: &str) -> Result<Region, ParseError> {
     let uri = endpoint.parse::<Uri>().context(EndpointParseError)?;
-    let name = region_name_from_host(uri.host().unwrap_or(""));
+    let name = region_name_from_host(uri.host().unwrap_or(""))
+        .unwrap_or_else(|| Region::default().name().into());
+    let endpoint = strip_endpoint(&uri);
+    Ok(Region::Custom { name, endpoint })
+}
 
-    // Reconstitute the endpoint from the URI, but strip off all path components
+/// Reconstitute the endpoint from the URI, but strip off all path components
+fn strip_endpoint(uri: &Uri) -> String {
     let pq_len = uri
         .path_and_query()
         .map(|pq| pq.as_str().len())
         .unwrap_or(1);
     let endpoint = uri.to_string();
-    let endpoint = endpoint[..endpoint.len() - pq_len].to_string();
-
-    Ok(Region::Custom { name, endpoint })
+    endpoint[..endpoint.len() - pq_len].to_string()
```

Comment:
Probably not a problem, but do we need to worry about multi-byte characters with the string indexing here?

Reply:
I don't see how. Both lengths are already proper indexes if the substrings contain multi-byte characters. This is just getting the byte index at which the hostname ends.